runtime/runtime/src/lib.rs

```diff
@@ -305,138 +378,163 @@ impl Runtime {
         gas_price: Balance,
-        signed_txs: impl IntoParallelIterator<Item = SignedTransaction>,
+        block_height: BlockHeight,
         current_protocol_version: ProtocolVersion,
-    ) -> Vec<(CryptoHash, Result<(ValidatedTransaction, TransactionCost), InvalidTxError>)> {
-        signed_txs
-            .into_par_iter()
-            .map(|signed_tx| {
-                (
-                    signed_tx.get_hash(),
-                    match validate_transaction(config, signed_tx, current_protocol_version) {
-                        Ok(validated_tx) => {
-                            match tx_cost(
-                                config,
-                                &validated_tx,
-                                gas_price,
-                                current_protocol_version,
-                            ) {
-                                Ok(cost) => Ok((validated_tx, cost)),
-                                Err(e) => Err(InvalidTxError::from(e)),
-                            }
-                        }
-                        Err((e, _tx)) => Err(e),
-                    },
-                )
-            })
-            .collect()
+        mut temp_state: Option<(Account, AccessKey)>,
+    ) -> Result<BatchResult, RuntimeError> {
+        let mut processed_transactions = Vec::with_capacity(batch.indices.len());
+
+        for (idx, transaction) in batch.iter() {
+            metrics::TRANSACTION_PROCESSED_TOTAL.inc();
+
+            let validated_tx = match validate_transaction(
+                &apply_state.config,
+                transaction.clone(),
+                current_protocol_version,
+            ) {
+                Ok(validated_tx) => validated_tx,
+                Err((e, tx)) => {
+                    Self::handle_invalid_transaction(
+                        e,
+                        &tx.get_hash(),
+                        current_protocol_version,
+                        "transaction validation failed",
+                    )?;
+                    continue;
+                }
+            };
+
+            let cost = match tx_cost(
+                &apply_state.config,
+                &validated_tx,
+                gas_price,
+                current_protocol_version,
+            ) {
+                Ok(cost) => cost,
+                Err(e) => {
+                    Self::handle_invalid_transaction(
+                        e,
+                        &validated_tx.get_hash(),
+                        current_protocol_version,
+                        "transaction cost calculation failed",
+                    )?;
+                    continue;
+                }
+            };
+
+            match verify_and_charge_tx_ephemeral(
+                &apply_state.config,
+                state_update,
+                &validated_tx,
+                &cost,
+                Some(block_height),
+                current_protocol_version,
+                temp_state.take(),
```

Comment:
I don't understand why you need to `take` here.

Reply:
Otherwise, I'll have to `.clone()`. It also ensures that we don’t accidentally reuse the same state for multiple txs